### ISSUE TITLE
Precompilation fails when STDOUT is an IOStream

### ISSUE THREAD

**Issue by garrison** [JuliaLang/julia#12975]
Consider the following very simple `test.jl`:

```julia
@show typeof(STDOUT)
using Compat
```

If run such that `STDOUT` is a tty, the precompilation succeeds.  However, if `STDOUT` is not a tty, it fails:

```sh
$ rm -f ~/.julia/lib/v0.4/Compat.ji
$ julia test.jl > /tmp/outfile
ERROR: LoadError: MethodError: `open` has no method matching open(::Cmd, ::ASCIIString, ::IOStream)
Closest candidates are:
  open(!Matched::Function, ::Any...)
  open(::Base.AbstractCmd, ::AbstractString)
  open(::Base.AbstractCmd, ::AbstractString, !Matched::Base.AsyncStream)
  ...
 in create_expr_cache at loading.jl:303
 in compilecache at loading.jl:349
 in require at ./loading.jl:217
 in include at ./boot.jl:260
 in include_from_node1 at ./loading.jl:271
 in process_options at ./client.jl:308
 in _start at ./client.jl:411
while loading /tmp/test.jl, in expression starting on line 5
```

Possibly (?) related: #12941

**Comment by vtjnash** [comment]
fixed by https://github.com/JuliaLang/julia/pull/12839

**Comment by garrison** [comment]
One annoying thing about this bug is that you can establish a workflow that seems to work reliably, but fails as soon as you run `Pkg.update()` and things need to be re-precompiled.  This could be an unwelcome surprise to anyone using julia in production (and directing stdout to a file) if this is not fixed before 0.4 final.

**Comment by garrison** [comment]
I'm going to add this to the 0.4.0 milestone because it is a regression in behavior from 0.3.  Feel free to move to a later milestone if you don't think this bug should hold up the release.

**Comment by garrison** [comment]
To be clear, this issue is fixed on master with #12839.  The remaining question is how to backport a fix for 0.4.  Since #12839 no longer renames `AsyncStream`, is it possible/nonbreaking to do a straightforward backport to 0.4?

**Comment by tkelman** [comment]
I'll try putting it on a backport branch (probably after rc3 to avoid pushing that back any further) and evaluate whether it breaks anything via pkgeval

**Comment by garrison** [comment]
Sounds like a great plan.  Thanks @tkelman!

**Comment by garrison** [comment]
Fixed in v0.4.0-rc4.
